lib/cretonne/src/isa/intel/mod.rs

```diff
@@ -111,4 +117,149 @@ impl TargetIsa for Isa {
     fn reloc_names(&self) -> &'static [&'static str] {
         &binemit::RELOC_NAMES
     }
+
+    fn prologue_epilogue(&self, func: &mut ir::Function) -> result::CtonResult {
+        let word_size = if self.flags().is_64bit() { 8 } else { 4 };
+        let csr_type = if self.flags().is_64bit() {
+            ir::types::I64
+        } else {
+            ir::types::I32
+        };
+        let csrs = abi::callee_saved_registers(&self.shared_flags);
+
+        let mut csr_stack_size = word_size; // Size of RBP to start with
+        for _reg in &csrs {
+            csr_stack_size += word_size;
+        }
+
+        let stack_offset = -(csr_stack_size as i32);
+        let slot = ir::StackSlotData {
+            kind: ir::StackSlotKind::IncomingArg,
+            size: csr_stack_size,
+            offset: stack_offset,
+        };
+        func.create_stack_slot(slot);
+
+        let total_stack_size = layout_stack(&mut func.stack_slots, word_size)?;
+        let local_stack_size = total_stack_size - csr_stack_size;
+
+        // Build up list of args, which we'll append forwards to the params and
+        // backwards to the returns.
+        let mut csr_args = Vec::new();
+        csr_args.push(ir::AbiParam::special_reg(
+            csr_type,
+            ir::ArgumentPurpose::FramePointer,
+            RU::rbp as RegUnit,
+        ));
+        for reg in &csrs {
+            csr_args.push(ir::AbiParam::special_reg(
+                csr_type,
+                ir::ArgumentPurpose::CalleeSaved,
+                *reg as RegUnit,
+            ));
+        }
+
+        for csr_arg in &csr_args {
+            func.signature.params.push(*csr_arg);
+        }
+        for csr_arg in csr_args.iter().rev() {
+            func.signature.returns.push(*csr_arg);
```

Comment:
I see what you're doing here, but to be clear, the order of these arguments only matter relative to the order of arguments on the return instruction.